### PR TITLE
Add AWS example for building a postgres RDS database with crossplane

### PR DESCRIPTION
**Pull request by mproffitt**
This PR introduces a complete step by step to delivering `postgresql` as an example for `crossplane` on `AWS`.

This includes a full end to end workflow, from setting up custom roles and polices for crossplane to assume, to deploying the providers, deploying the composition, making a claim against it and finally connecting to the database from within the cluster.

### CODE REVIEW

**Comment by mproffitt** [comment]
There's a couple of things I'm wondering about with the documentation here.

1. The first section in [`README.md`](https://github.com/giantswarm/crossplane-examples/tree/add-aws-crossplane/aws-provider/postgresdb#create-the-primary-role) covers setting up a `WebIdentityAssumeRole` for the crossplane service accounts to use. What I'm unsure of here is exactly where is this information best served from? We have similar information in the Azure folder ([see here](https://github.com/giantswarm/crossplane-examples/blob/add-aws-crossplane/azure-provider/postgresdb/README.md#azure-setup)) but that doesn't really include the pod annotations for workload identity which is something I think is worth capturing, nor does the example use it.
2. The discussion in section 2, setting up the secondary account could be more explicit but I wanted to avoid the duplication of information. How would this section need to read if we split out the first section?
3. The section on [enabling the providers](https://github.com/giantswarm/crossplane-examples/tree/add-aws-crossplane/aws-provider/postgresdb#enabling-providers) is almost certainly better served from [management-cluster-bases](https://github.com/giantswarm/management-cluster-bases/tree/main/extras/crossplane/providers) from where it will be consumed. In fact this information is already partially duplicated to that location.

Regarding the cloud architecture; This sets up a private VPC for the database with 3 subnets starting in the middle of the `10.1.0.0/25` range which is above the cluster cidr to avoid conflicts.

Why 3 subnets?

To answer that we need to look at the SubnetGroup for the database. This was required to get the RDS assigned into the subnets for the VPC and not assigned into the `default` subnet group. `SubnetGroup` objects require a minimum of 2 subnets and creating 3 means there would be room to expand the database to 2 read replicas, one in each of the availability zone `B` and `C`.

The drawback to this approach is the number of components then required which is:

1 x VPC
1 x SecurityGroup
1 x SecurityGroupRule
1 x PeeringConnection
3 x Subnet
3 x RouteTable
3 x RouteTableAttachments
6 x Routes (3 in each VPC, attached to the private subnets of the cluster)
1 x KMS key for Database encryption
1 x RDS instance

~~I'm not entirely convinced I've got all the naming and tagging correct and this may need re-visiting for correctness, completeness and polish~~.

File: aws-provider/postgresdb/README.md [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+management cluster, then using that to bootstrap a database using a secondary
+role assumed by crossplane for the actual build.
+
+> **Warning** Creating IAM permissions
+>
+> The  following document assumes that you have permissions to create IAM roles
+> and policies inside your account.
+>
+> If you do not have such permissions, please reach out to your account
+> administrator and ask them to create these for you.
```
Comment (pipo02mix) [inline aws-provider/postgresdb/README.md]:
```suggestion
__Warning__: The following document assumes that you have permission to create IAM roles and policies inside your account. If you do not have such permissions, please reach out to your account administrator and ask them to create these for you.
```
Following the same syntaxis as our docs?

Comment (pipo02mix) [reply aws-provider/postgresdb/README.md]: Checking the other clarification blocks you have below, I tend to think using quote block it is better when is a multi-paragraph block, so ignore me

File: aws-provider/postgresdb/README.md [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+> {
+>     "Sid": "VisualEditor2",
+>     "Effect": "Allow",
+>     "Action": "sts:AssumeRole",
+>     "Resource": "*",
+>     "Condition": {
+>         "ForAnyValue:StringLike": {
+>             "aws:PrincipalArn": [
+>                 "arn:aws:sts::${AWS_ACCOUNT_ID}:role/crossplane-assume-role",
+>                 "arn:aws:sts::${AWS_ACCOUNT_ID}:assumed-role/rds-crossplane-role*"
```
Comment (pipo02mix) [inline aws-provider/postgresdb/README.md]: TIL you need `assumed-role`

File: aws-provider/postgresdb/README.md [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+│       └── kustomization.yaml
+```
+
+into each `kustomization.yaml` file, copy the following contents:
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+resources:
+  - https://github.com/giantswarm/management-cluster-bases//extras/crossplane/providers/upbound/aws?ref=main
```
Comment (pipo02mix) [inline aws-provider/postgresdb/README.md]: This gives me 404, should be `https://github.com/giantswarm/management-cluster-bases/tree/main/extras/crossplane/providers/upbound/aws?ref=main` ?

File: aws-provider/postgresdb/README.md [hunk 4]
```diff
@@ -0,0 +1,346 @@
+# `crossplane` PostgresDB example
+
+In this more complex example, we're going to walk through the process of
+enabling provider-families with `sts:AssumeRoleWithWebIdentity` to the
+management cluster, then using that to bootstrap a database using a secondary
+role assumed by crossplane for the actual build.
```
Comment (pipo02mix) [inline aws-provider/postgresdb/README.md]: When you mention "more complex example" you compare with...?  maybe a link to the simple one (just in case someone arrives here directly)

File: aws-provider/postgresdb/README.md [hunk 5]
```diff
@@ -0,0 +1,346 @@
+# `crossplane` PostgresDB example
+
```
Comment (pipo02mix) [inline aws-provider/postgresdb/README.md:2]: I would add a brief description first of the goal of this entire tutorial. Something along the lines "We want to deploy a RDS Postgres database using Crossplane ... " before you jump into the flesh

File: aws-provider/postgresdb/README.md (hunk 4 above)
Comment (pipo02mix) [reply aws-provider/postgresdb/README.md]: and maybe add why we suggest this approach (segment role permissions...)

File: aws-provider/postgresdb/README.md [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+> **Warning** Creating IAM permissions
+>
+> The  following document assumes that you have permissions to create IAM roles
+> and policies inside your account.
+>
+> If you do not have such permissions, please reach out to your account
+> administrator and ask them to create these for you.
+
+## Create the primary role
+
```
Comment (pipo02mix) [inline aws-provider/postgresdb/README.md]: Brief intro? like "We are creating a role `crossplane-assume-role` for crossplane controller to let it assume the specific provider families roles where the permission for RDS are... "

File: aws-provider/postgresdb/README.md [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+       aws iam attach-role-policy --policy-arn \
+           "arn:aws:iam::aws:policy/crossplane-assume-role-policy" \
+           --role-name crossplane-assume-role
+     ```
+
+     If the role already exists, update its trust policy with
+
+     ```bash
+       aws iam update-assume-role-policy --role-name crossplane-assume-role \
+           --policy-document file://policies/crossplane-assume-role-trust-policy.json
```
Comment (pipo02mix) [inline aws-provider/postgresdb/README.md]: is this action additive? or you need to add existing permissions to `crossplane-assume-role-trust-policy.json` first?

File: aws-provider/postgresdb/README.md [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+>
+> Without this, crossplane may not be able to assume the role and you will
+> experience permissions failures when building resources.
+
+The policy file [rds-crossplane-policy](./policies/rds-crossplane-policy.json)
+is a restrictive policy that allows only the specific actions required to build
+this example.
+
+## Enabling providers
+
```
Comment (pipo02mix) [inline aws-provider/postgresdb/README.md]: Maybe a short intro, "Since new crossplane versions come without any API enabled, we need to add every AWS family/service to Crossplane in order to...."

**Comment by pipo02mix** [comment]
> The first section in [README.md](https://github.com/giantswarm/crossplane-examples/tree/add-aws-crossplane/aws-provider/postgresdb#create-the-primary-role) covers setting up a WebIdentityAssumeRole for the crossplane service accounts to use. What I'm unsure of here is exactly where is this information best served from? We have similar information in the Azure folder ([see here](https://github.com/giantswarm/crossplane-examples/blob/add-aws-crossplane/azure-provider/postgresdb/README.md#azure-setup)) but that doesn't really include the pod annotations for workload identity which is something I think is worth capturing, nor does the example use it

Yeah I have been thinking about this. IMO it would be great to not duplicate info and maybe do a single doc about setting permission which refers to both guides?

File: aws-provider/postgresdb/README.md (hunk 3 above)
Comment (mproffitt) [reply aws-provider/postgresdb/README.md]: No because this is a command to `kustomize` which is copying the remote refs (See [kustomize examples](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/remoteBuild.md#examples))

* With `tree/main` included:
   ```
   $ kustomize build .
   Error: accumulating resources: accumulating resources from 
   'https://github.com/giantswarm/management-cluster-bases/tree/main/extras/crossplane/providers/upbound/aws?ref=main': 
   MalformedYAMLError: yaml: line 203: mapping values are not allowed in this context in File: 
   https://github.com/giantswarm/management-cluster-bases/tree/main/extras/crossplane/providers/upbound/aws?ref=main
   ```
   
 * Without:
    ```
   $ kustomize build .
   apiVersion: v1
   kind: ServiceAccount
   metadata:
     annotations:
   ...
    ```

File: aws-provider/postgresdb/README.md (hunk 4 above)
Comment (mproffitt) [reply aws-provider/postgresdb/README.md]: Maybe I need to give some additional thought to the wording in this section. All of the AWS examples so far are "simple" in that none of them create more than a single piece of infrastructure and I made the assumption that because the repo is `crossplane-examples` users would look at this as a whole.

Let me consider how to rephrase this and update :small_airplane:

File: aws-provider/postgresdb/README.md (hunk 7 above)
Comment (mproffitt) [reply aws-provider/postgresdb/README.md]: This is already highlighted that there is an "either or" operation taking place here. This is captured in a sub-bullet as follows

> * Check that the role does not already exist
>   If it does not exist do this
>   If it exists do this

As far as I'm aware, it's possible to run both of these commands and the update following create would just be a NOOP. The command effectively replaces the trust policy with the policy you give it.

File: aws-provider/postgresdb/README.md (hunk 8 above)
Comment (mproffitt) [reply aws-provider/postgresdb/README.md]: I would disagree with this wording. New crossplane versions do not have a concept of an API and don't remove providers; similarly, the wording "We need to add every AWS family/service to Crossplane..." would imply that we're going to install every family member, all 159 of them.

This would offer even worse performance to the cluster than just installing `provider-aws` as this would pull in 159 additional deployments, each scalable to `n` replicas, this would be on top of the 903 CRDs it would also be installing.

Maybe something closer to this?

> By default, crossplane comes without any providers enabled and as such, we need to install the providers we're going to use for building and managing our infrastructure.
> 
> In this exampled we require the following providers to be enabled.
> 
> - EC2
> - RDS
> - KMS
> 
> The name of each provider may be discovered from its API group prefixed with `xpkg.upbound.io/provider-aws-`. 
> 
> For example, for `ec2.aws.upbound.io` becomes `xpkg.upbound.io/provider-aws-ec2@VERSION` where `VERSION` is the exact provider version you wish to install, discoverable from [doc.crds.dev/github.com/upbound/provider-aws](https://doc.crds.dev/github.com/upbound/provider-aws)

Comment (pipo02mix) [reply aws-provider/postgresdb/README.md]: that sounds good to me ;)

**Review by kubasobon — APPROVED** [review]

File: aws-provider/postgresdb/README.md [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+For our build to take place, we require the creation of two roles in AWS IAM.
+
+The first of these is an `AssumeRoleWithWebIdentity` which has permissions to
+assume other roles, either in the current account, or using cross-account
+permissions.
+
+For this example, we're going to assume both roles are encompassed in the same
+account.
+`
+1. Edit the file [`crossplane-assume-role-policy`](./policies/crossplane-assume-role-policy.json)
```
Comment (piontec) [inline aws-provider/postgresdb/README.md]: Maybe let's add a comment, that this policy lists all IAM roles you're allowing to claim?

**Review by piontec — COMMENTED** [review]
Great and thanks for writing it down!

A few remarks:
1. Ugh, the AWS IAM setup is pretty hard to follow - by reading it briefly, without checking how everything works, I still didn't get it. Maybe you could start with a picture of objects involved in the security setup and how they relate to each other?
2. Splitting the doc. I would split it into 3 parts:
    1. A general explanation (non MCB specific) about how to setup IRSA+Web Identity auth
    2. MCB specific patches and stuff that make the deployment easier (and probably move that closer/directly to the MCB repo)
    3. Actual provider and objects required to create an RDS

With a split like above, the 3rd part shouldn't be too long, and I like the fact that we also have a more comprehensive example.

**Comment by mproffitt** [comment]
I have restructured a lot of this example and split this into a number of separate documents over two repositories

- Provider setup and Primary role go to `management-cluster-bases` as part of this PR https://github.com/giantswarm/management-cluster-bases/pull/29
- Secondary role setup and RDS deployment remain as part of this repository
- A breakdown of IAM permissions required by this build can be found in the document [`policies/README.md`](https://github.com/giantswarm/crossplane-examples/blob/a404d05cc60c68b4ea00b8a0a06ab0c8947583d0/aws-provider/postgresdb/policies/README.md)

**Review by pipo02mix — APPROVED** [review]
That's much better. LGTM if piontec is happy

File: aws-provider/postgresdb/README.md [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+Before this example can be built, you must first enable crossplane AWS provider
+family members and create the primary role for crossplane to assume.
+
+In this example, we need the following provider family members enabled inside the
+cluster.
+
+- [provider-aws-ec2](https://marketplace.upbound.io/providers/upbound/provider-aws-ec2/)
+- [provider-aws-rds](https://marketplace.upbound.io/providers/upbound/provider-aws-rds/)
+- [provider-aws-kms](https://marketplace.upbound.io/providers/upbound/provider-aws-ec2/)
```
Comment (uvegla) [inline aws-provider/postgresdb/README.md]:
```suggestion
- [provider-aws-kms](https://marketplace.upbound.io/providers/upbound/provider-aws-kms/)
```

File: aws-provider/postgresdb/README.md [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+1. Edit the file [`rds-crossplane-policy](./policies/rds-crossplane-policy.json)
+   and set the variable `${AWS_ACCOUNT_ID}` to the ID of the AWS account you're
+   going to use.
+
+   ```bash
+   export POLICY_NAME="rds-crossplane-policy"
+   export AWS_ACCOUNT_ID=$(aws sts get-caller-identity | jq -rn .Account)
+
+   cat ./policies/${POLICY_NAME}.json | envsubst \
+       | sponge ./policies/${POLICY_NAME}.json
```
Comment (uvegla) [inline aws-provider/postgresdb/README.md:48]: Why not just redirect the output to the file? I dont really know `sponge`, just quickly giving it a try it is equivalent to:

```bash
cat ./policies/${POLICY_NAME}.json | envsubst > ./policies/${POLICY_NAME}.json
```

Just so that it is one less tool needed to follow the example. `envsubst` is not on mac either by default, but needed.

**Review by uvegla — APPROVED** [review]

File: aws-provider/postgresdb/README.md (hunk 11 above)
Comment (mproffitt) [reply aws-provider/postgresdb/README.md:48]: I did this for some level of consitency between the handling of outputs of tooling. 

`jq` does not allow for `--inplace` or `-i` functionality in the same way that `yq` and `sed` have which, if means `jq file.json > file.json` actually forces truncation of the file.

Whilst a better argument here would be "Why not use `yq` for handling JSON", I feel this drives some level of confusion when using a tool designed for yaml against json policies.

To maintain a sense of similarity between the command outputs I re-used the pipe to `sponge` even though this is a redundant requirement.

Regarding `envsubst` this is called out already on the [`README.md`](https://github.com/giantswarm/crossplane-examples/blob/main/README.md) for the repository as a requirement.
